doc/developer: update and restructure guide (#12199)
MaterializeInc/materialize: doc/developer/guide.md

    entry. Changes to Materialize very rarely require changes in rust-dec, so
    maintaining the two separately does not introduce much overhead.

## Developer tools

### Editors and IDEs

> In principle, any text editor can be used to edit Rust code. There is nothing
> special about Materialize's codebase: any general Rust development setup should
> work fine.

Reviewer: I wrote this paragraph, but I don't quite agree with it anymore. Materialize now _is_ more complicated than a typical Rust project, in that you can't just run `cargo run` or `cargo test` and expect things to work.

I think we can leave in the bit about any text editor being fine, but delete the (sadly no longer true) bit about how Mz is just a vanilla Rust binary.

```suggestion
In principle, any text editor can be used to edit Rust code, including Materialize.
```

Author: Ack, done!
